### PR TITLE
Qpid Reconnect to Broker

### PR DESCRIPTION
**Pull request by nguyenfilip**

### CODE REVIEW

File: server/src/main/java/org/candlepin/audit/QpidConnection.java [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+
+    /**
+     * This indempotent method will establish connection to qpid broker. Per JMS standard
+     * it must recreate all JMS objects such as Connection, TopicSession, TopicPublisher.
+     * @throws Exception errors during connecting to the Broker
+     */
+    public void connect() throws Exception {
+        synchronized (connectionLock) {
+            if (connectionStatus == STATUS.CONNECTED) {
+                log.info("Already connected to Qpid Broker, ignoring this connect request");
```
Comment (dsarman) [inline server/src/main/java/org/candlepin/audit/QpidConnection.java]: Are we really ignoring the request? The method seems to go on normally except for the log message.

File: server/src/main/java/org/candlepin/config/ConfigProperties.java [hunk 2]
```diff
@@ -336,8 +336,8 @@ private ConfigProperties() {
             this.put(AMQP_TRUSTSTORE, "/etc/candlepin/certs/amqp/candlepin.truststore");
             this.put(AMQP_TRUSTSTORE_PASSWORD, "password");
 
-            this.put(AMQP_CONNECTION_RETRY_INTERVAL, "10"); // Every 10 seconds
-            this.put(AMQP_CONNECTION_RETRY_ATTEMPTS, "12"); // Try for 2 mins (10s * 12)
+            this.put(AMQP_CONNECTION_RETRY_INTERVAL, "10"); // Every 5 seconds
```
Comment (dsarman) [inline server/src/main/java/org/candlepin/config/ConfigProperties.java]: Is the value 10 correct? According to comment it should be 5.

File: server/src/main/java/org/candlepin/audit/QpidConnection.java (hunk 1 above)
Comment (nguyenfilip) [inline server/src/main/java/org/candlepin/audit/QpidConnection.java]: Nice catch. Will try to add some test for this

**Comment by nguyenfilip** [comment]
retest this please

**Comment by nguyenfilip** [comment]
retest this please

**Comment by nguyenfilip** [comment]
retest this please

**Comment by mstead** [comment]
One of the issues that I've seen in the past (that I don't **think** we had a solution for) is that when a time out occurs and all connection retries have happened, an exception is thrown by Qpid and there is no way to get the connection back, short of a tomcat restart. Does this PR address this as well?

**Comment by mstead** [comment]
> @nguyenfilip One of the issues that I've seen in the past (that I don't think we had a solution for) is that when a time out occurs and all connection retries have happened, an exception is thrown by Qpid and there is no way to get the connection back, short of a tomcat restart. Does this PR address this as well?

I think this might be handled in another one of your PRs for QPID.

File: server/src/main/java/org/candlepin/audit/QpidConnection.java (hunk 1 above)
Comment (nguyenfilip) [inline server/src/main/java/org/candlepin/audit/QpidConnection.java]: I will remove this check all together. I found that its better to just let the reconnect to go through. The connectionStatus variable value is not reliable in concurrent environment.

File: server/src/main/java/org/candlepin/audit/QpidConnection.java [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+/**
+ * Connection to a Qpid Broker. It can be in two states: CONNECTED or DOWN.
+ * The class has ability to reconnect to the broker.
+ *
+ * This class also orchestrates all the initial configuration of the
+ * Qpid connection
+ * @author fnguyen
+ *
+ */
+public class QpidConnection {
```
Comment (mstead) [inline server/src/main/java/org/candlepin/audit/QpidConnection.java]: Perhaps javadoc the public methods for this class.

Comment (nguyenfilip) [inline server/src/main/java/org/candlepin/audit/QpidConnection.java]: thanks, added

Comment (mstead) [inline server/src/main/java/org/candlepin/audit/QpidConnection.java]: :+1:

File: server/src/main/java/org/candlepin/audit/QpidConnection.java [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+        }
+
+    }
+
+    public Connection newConnection() throws JMSException {
+        log.debug("creating connection");
+        Connection conn = null;
+        conn = connectionFactory.createConnection();
+        conn.start();
+        return conn;
```
Comment (nguyenfilip) [inline server/src/main/java/org/candlepin/audit/QpidConnection.java]: I don't think this is a good idea. I would rather have false positive (connectionStatus = CONNECTED) then having DOWN incorrectly.

This is kind of tricky to get right, multiple threads might try to reconnect and I would rather set connectionStatus = DOWN only when I am absolutely sure the send failed.

**Comment by nguyenfilip** [comment]
@mstead I think this can go in if you don't have more concerns

File: server/src/main/java/org/candlepin/audit/QpidConnection.java (hunk 4 above)
Comment (mstead) [inline server/src/main/java/org/candlepin/audit/QpidConnection.java]: Ok, makes sense.

**Comment by mstead** [comment]
> @mstead I think this can go in if you don't have more concerns

Looks good to me.

**Pull request merged**